copy/copy.go

```diff
@@ -655,17 +725,122 @@ func (ic *imageCopier) copyLayer(ctx context.Context, srcInfo types.BlobInfo, po
 	cachedDiffID := ic.c.blobInfoCache.UncompressedDigest(srcInfo.Digest) // May be ""
 	diffIDIsNeeded := ic.diffIDsAreNeeded && cachedDiffID == ""
 
+	// Create the initial placeholder bar which is later being replaced by
+	// another one.
+	bar := ic.c.createProgressBar(pool, srcInfo, "blob", " ", mpb.BarRemoveOnComplete())
+
 	// If we already have the blob, and we don't need to compute the diffID, then we don't need to read it from the source.
 	if !diffIDIsNeeded {
-		reused, blobInfo, err := ic.c.dest.TryReusingBlob(ctx, srcInfo, ic.c.blobInfoCache, ic.canSubstituteBlobs)
+		// tryReusingBlob is an anonymous function to avoid adding
+		// another method with a complex parameter list and semantics.
+		// It's sole purpose is to simplify the control with an early
+		// return if the image destination does not blob locking.
+		tryReusingBlob := func() (bool, types.BlobInfo, digest.Digest, error) {
+			reused, blobInfo, err := ic.c.dest.TryReusingBlob(ctx, srcInfo, ic.c.blobInfoCache, ic.canSubstituteBlobs)
+			if err != nil {
+				return false, types.BlobInfo{}, "", errors.Wrapf(err, "Error trying to reuse blob %s at destination", srcInfo.Digest)
+			}
+			if reused {
+				logrus.Debugf("Skipping blob %s (already present):", srcInfo.Digest)
+				bar.SetTotal(0, true)
+				ic.c.createProgressBar(pool, srcInfo, "blob", "skipped: already exists", mpb.BarReplaceOnComplete(bar))
+				return true, blobInfo, cachedDiffID, nil
+			}
+
+			if !ic.c.dest.SupportsBlobLocks() {
```

Comment:
This is too big for a single (nested!) function.

Maybe extract the code, starting roughly from this line, until the `!copiedByAnotherProcess` into a separate function? The logic to block and wait seems fairly self-contained.

---

Or, well, I’d still rather prefer for all of this to go into `storageImageDestination.TryReusingBlob`, so that the locking/unlocking logic is localized to that transport.

Reply:
Done.